### PR TITLE
Stride `MatmulOp` according to set allocation domain

### PR DESCRIPTION
**Pull request by Priya2698**
Resolves Issue #2427.
If the `MatmulOp` has a stride order set from python frontend (`fd.ops.add_output/fd.ops.stride_order`), returns a copy of the output with the specified memory_layout.

`at::matmul_out` is not used since it does not allow inputs/outputs which require gradients.
https://github.com/pytorch/pytorch/blob/1f3d8896bc9cea7f46c50ff92b69c6aa139defcb/aten/src/ATen/native/LinearAlgebra.cpp#L2018-L2025

### CODE REVIEW

**Comment by Priya2698** [comment]
!test

**Comment by Priya2698** [comment]
!test

File: csrc/ir/nodes.cpp [hunk 1]
```diff
@@ -4355,7 +4356,14 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
     const std::vector<PolymorphicValue>& inputs) const {
   const auto a = inputs.at(0).as<at::Tensor>();
   const auto b = inputs.at(1).as<at::Tensor>();
-  return {at::matmul(a, b)};
+  auto matmul_out = at::matmul(a, b);
+  if (out()->hasAllocation()) {
+    auto matmul_sizes = matmul_out.sizes().vec();
```
Comment (wujingyue) [inline csrc/ir/nodes.cpp:4361]: How about changing computeStrides to take c10::IntArrayRef? `.vec()` copies.

File: csrc/ir/nodes.cpp [hunk 2]
```diff
@@ -4355,7 +4356,14 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
...
+    auto strides = computeStrides(out(), matmul_sizes);
+    matmul_out = at::as_strided(matmul_out, matmul_sizes, strides);
```
Comment (wujingyue) [inline csrc/ir/nodes.cpp:4363]: This feels wrong. IIUC, as_strided creates a view of the input. So if the input tensor is in the wrong memory format there's no way to "relayout" the storage using as_strided. I think you would have caught this problem if [your test](https://github.com/NVIDIA/Fuser/pull/3447/files#diff-22fe4d3f2193276335bba1281b92e245e9ed43e4631a1f17bad63d2cfca0119bR224) had also verified the content (not just the shape) of the output tensor.

File: csrc/ir/nodes.cpp [hunk 3]
```diff
@@ -4355,7 +4356,14 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
...
+  }
+  inferAndValidateAllocationSizesAndStrides(matmul_out, out(), ee);
```
Comment (wujingyue) [inline csrc/ir/nodes.cpp:4365]: I'm not sure about validating output allocation for all MatmulOps. 
1. We already validate allocation sizes/strides for each segment's [inputs](https://github.com/NVIDIA/Fuser/blob/c154e90919c40bfe2202b432c2a38e106d1a5444/csrc/tensor_metadata.cpp#L365) and [outputs](https://github.com/NVIDIA/Fuser/blob/c154e90919c40bfe2202b432c2a38e106d1a5444/csrc/runtime/allocations.cpp#L307). Given MatmulOp currently forms its own segment, existing validation seems enough. 
2. If/When MatmulOp produces an **internal** tensor, we can't always materialize the tensor as an `at::Tensor` that matches its allocation domain. For example, the allocation domain can be a split and/or a swizzle of logical. Assuming allocation is a permutation of logical is probably OK for segment inputs/outputs, but can be too limiting for internal tensors. cc @zasdfgbnm

File: csrc/ir/utils.h [hunk 4]
```diff
@@ -757,4 +757,8 @@ std::vector<IterDomain*> strideOrderToAllocation(
     const std::vector<IterDomain*>& logical_domain,
     const std::vector<int64_t>& stride_order);
 
+std::vector<int64_t> computeStrides(
```
Comment (jjsjann123) [inline csrc/ir/utils.h:760]: looks like we forgot this declaration?

File: tests/python/test_matmul.py [hunk 5]
```diff
@@ -201,3 +201,24 @@ def fusion_func(fd: FusionDefinition) -> None:
...
+                b = fd.from_pytorch(inputs[1])
+                out = fd.ops.matmul(a, b)
+                fd.add_output(out, stride_order=perm)
+
+            with FusionDefinition() as fd:
+                fusion_func(fd)
+            out = fd.execute(inputs)
+            verify_stride_order(out[0].stride(), perm)
+            # Verify that setting the stride order does not change the logical shape
+            self.assertEqual(out[0].shape, torch.Size([b, b, m, n]))
```
Comment (jjsjann123) [inline tests/python/test_matmul.py:224]: You forgot to validate the result here. That's probably why the issue @wujingyue raised didn't get caught.

File: csrc/ir/nodes.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/ir/nodes.cpp:4363]: Agreed! you need an explicit copy here instead.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]

File: csrc/ir/nodes.cpp (hunk 2 above)
Comment (wujingyue) [reply csrc/ir/nodes.cpp:4363]: There may be a better way than copy. I suspect matmul_out respects the output strides, but please double check

Comment (jjsjann123) [reply csrc/ir/nodes.cpp:4363]: Good call. if we can feed our own outputs there, we might as well do that. So you need to call `as_strided` on output before executing the kernel.

BTW, aten may or may not do that in a fused kernel. i.e. https://github.com/pytorch/pytorch/blob/1f3d8896bc9cea7f46c50ff92b69c6aa139defcb/aten/src/ATen/native/LinearAlgebra.cpp#L2097-L2099

I think the more interesting question is, do we assume it's always a copy and should we try to expose this transpose/copy kernel in fusion for nvfuser to handle it instead... :thinking:

Comment (wujingyue) [reply csrc/ir/nodes.cpp:4363]: > I think the more interesting question is, do we assume it's always a copy and should we try to expose this transpose/copy kernel in fusion for nvfuser to handle it instead...

It's hard to be performance optimal given the implementation of at::matmul_out can change without notifying us. I'd check for the cases that we care about. I suspect for all or 90% of them at::matmul_out is able to fuse. If so, I'd unconditionally assume it'll fuse for simplicity.

Comment (Priya2698) [reply csrc/ir/nodes.cpp:4363]: Ahh you're right. I misunderstood this function call -- there is no copy here. Will fix this to use matmul_out or an explicit copy.

**Comment by Priya2698** [comment]
!test

File: csrc/tensor_metadata.h [hunk 6]
```diff
@@ -115,4 +115,9 @@ inferAndValidateAllocationSizesAndStrides(
     TensorView* tv,
     ExpressionEvaluator ee);
 
+// Given a tensorview, compute the strides according to the allocation domain
+// for re-striding the corresponding ATen tensor.
+std::vector<int64_t> computeStrides(
```
Comment (wujingyue) [inline csrc/tensor_metadata.h:120]: Nit: I smell a YAGNI -- I think it's premature to expose this via a header whereas the only use is in tensor_metadata.cpp. I'd instead leave this helper function in tensor_metadata.cpp.

File: tests/python/test_matmul.py [hunk 7]
```diff
@@ -201,3 +201,24 @@ def fusion_func(fd: FusionDefinition) -> None:
         ]
         outputs, _ = self.exec_nvfuser(fusion_func, inputs)
         assert outputs[0].ndim == 3
+
+    def test_matmul_stride(self):
+        b, m, n, k = 3, 2, 5, 4
+        inputs = [
+            torch.randn(b, b, m, k, device="cuda", dtype=torch.float16),
```
Comment (wujingyue) [inline tests/python/test_matmul.py:208]: It's valid but weird to have two batch dimensions in input `a`. Would you mind making this more realistic?

File: tests/python/test_matmul.py [hunk 8]
```diff
@@ -201,3 +201,24 @@ def fusion_func(fd: FusionDefinition) -> None:
...
+            torch.randn(k, n, device="cuda", dtype=torch.float16),
+        ]
+        for perm in itertools.permutations(range(4), 4):
+
+            def fusion_func(fd: FusionDefinition) -> None:
+                a = fd.from_pytorch(inputs[0])
+                b = fd.from_pytorch(inputs[1])
```
Comment (wujingyue) [inline tests/python/test_matmul.py:215]: I don't like overloading `b` for both the batch size and the operand. I'd rename `a` to `x` and `b` to `y` for readability.

File: csrc/ir/nodes.cpp [hunk 9]
```diff
@@ -4371,7 +4372,17 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
     const std::vector<PolymorphicValue>& inputs) const {
   const auto a = inputs.at(0).as<at::Tensor>();
   const auto b = inputs.at(1).as<at::Tensor>();
-  return {at::matmul(a, b)};
+
+  auto matmul_out = at::matmul(a, b);
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp:4376]: at::matmul_out is not used since it does not allow inputs/outputs which require gradients.
https://github.com/pytorch/pytorch/blob/1f3d8896bc9cea7f46c50ff92b69c6aa139defcb/aten/src/ATen/native/LinearAlgebra.cpp#L2018-L2025

File: csrc/tensor_metadata.h (hunk 6 above)
Comment (Priya2698) [reply csrc/tensor_metadata.h:120]: I am using it in `nodes.cpp`.

File: tests/python/test_matmul.py (hunk 7 above)
Comment (Priya2698) [reply tests/python/test_matmul.py]: Switched it to be of sdpa pattern.

File: csrc/ir/nodes.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (wujingyue) [reply csrc/ir/nodes.cpp:4376]: This is suspicious -- We are not using ExpressionEvaluator to build a DAG for autograd, so inputs/outputs here shouldn't require grads. Where did inputs/outputs get requires_grads? Your test case didn't `torch.randn(..., requires_grad=True)` obviously to start with. 

cc @jjsjann123

File: csrc/tensor_metadata.h (hunk 6 above)
Comment (wujingyue) [reply csrc/tensor_metadata.h:120]: Sorry for missing the fact that computeStrides is used in nodes.cpp. However, the same thing -- it's only used in one source location so I'd rather it be defined in the same file (i.e. nodes.cpp). (It **could** be reused elsewhere but YAGNI)

File: csrc/ir/nodes.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp:4376]: You might be right, the tensor evaluations themselves will likely not have the requires_grad field.
I had ruled out `at::matmul_out` after going through its code, and trying it independently that raised this condition, which may have been premature.

[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp:4376]: I attempted a complete example in nvfuser and I do get an error.
I need to dig into the expression evaluator on how this flag is propagated/inferred. Expression evaluator will have this information for the fusion inputs.

File: tests/python/test_matmul.py [hunk 10]
```diff
@@ -201,3 +201,29 @@ def fusion_func(fd: FusionDefinition) -> None:
...
+                n, h, s, e, device="cuda", dtype=torch.float16, requires_grad=True
+            ),
+        ]
+        for perm in itertools.permutations(range(4), 4):
+
+            def fusion_func(fd: FusionDefinition) -> None:
+                q = fd.from_pytorch(inputs[0])
+                k = fd.from_pytorch(inputs[1])
+                kt = fd.ops.permute(k, [0, 1, 3, 2])
+                out = fd.ops.matmul(q, kt)
```
Comment (wujingyue) [inline tests/python/test_matmul.py:221]:
```suggestion
                k_t = fd.ops.permute(k, [0, 1, 3, 2])
                out = fd.ops.matmul(q, k_t)
```

Nit: `k_t` expresses better that it's a "transposition" of `k`.

**Review by wujingyue — APPROVED** [review]

**Comment by Priya2698** [comment]
!test

File: csrc/ir/nodes.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (wujingyue) [reply csrc/ir/nodes.cpp:4408]: I'm fine with addressing this in a separate PR. But I'd still try to understand the requires_grad bit sooner than later -- it shouldn't have been there and it blocks optimization to matmul into a pre-allocated output.

**Comment by Priya2698** [comment]
!test

File: csrc/ir/nodes.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/ir/nodes.cpp:4462]: Maybe we can try this https://github.com/pytorch/pytorch/blob/a3688ead4bcd13de7d81a9e114b1c8cfc332970a/c10/core/GradMode.h#L31-L33

[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/ir/nodes.cpp:4462]: but yeah, we can do it in a separate PR. I'm more than happy to review that.

**Review by jjsjann123 — APPROVED** [review]
Let's have a follow up PR switching to the inplace version!
